From patchwork Thu Feb  6 09:45:07 2014
Content-Type: text/plain; charset="utf-8"
MIME-Version: 1.0
Content-Transfer-Encoding: 7bit
X-Patchwork-Submitter: David Jedelsky <david.jedelsky@gmail.com>
X-Patchwork-Id: 22003
Received: from mail.tu-berlin.de ([130.149.7.33])
	by www.linuxtv.org with esmtp (Exim 4.72)
	(envelope-from <linux-media-owner@vger.kernel.org>)
	id 1WBLXi-0005VY-DL; Thu, 06 Feb 2014 10:46:46 +0100
X-tubIT-Incoming-IP: 209.132.180.67
Received: from vger.kernel.org ([209.132.180.67])
	by mail.tu-berlin.de (exim-4.72/mailfrontend-5) with esmtp
	id 1WBLXg-0002Rb-7U; Thu, 06 Feb 2014 10:46:46 +0100
Received: (majordomo@vger.kernel.org) by vger.kernel.org via listexpand
	id S1755899AbaBFJql (ORCPT <rfc822;mkrufky@linuxtv.org> + 1 other);
	Thu, 6 Feb 2014 04:46:41 -0500
Received: from mail-ea0-f172.google.com ([209.85.215.172]:54458 "EHLO
	mail-ea0-f172.google.com" rhost-flags-OK-OK-OK-OK) by vger.kernel.org
	with ESMTP id S1755374AbaBFJqh (ORCPT
	<rfc822; linux-media@vger.kernel.org>); Thu, 6 Feb 2014 04:46:37 -0500
Received: by mail-ea0-f172.google.com with SMTP id l9so650291eaj.3
	for <linux-media@vger.kernel.org>;
	Thu, 06 Feb 2014 01:46:36 -0800 (PST)
DKIM-Signature: v=1; a=rsa-sha256; c=relaxed/relaxed;
	d=gmail.com; s=20120113;
	h=from:to:cc:subject:date:message-id;
	bh=2BuI62NGhD3QrDqlArK0gXwPJ6z88AkAWtoNxBKLA5o=;
	b=oL3CvafHbMqU7lNWKQ21IGvn0QCk5KpK+D8BfuekKWSdWA1LT9MRuyxYtOZSiq4+g2
	/zHlt9gVtawIARppHfWLZFzvAXnOlbcSQtQxBmK/Bt1bmzazevcaLrk8uvDy1LdeR9A5
	bL8btnZoUp6aK0MRXKitMvA6uV6PnkeyMm0Xz+Xk+FZGNqKgBWgiRTjQcD6ANKcpZDzJ
	wxgnbTPHjLGGd4l8fr1Q5QLSmiD7A1IivYyEcr88OoNEkJiR4NSGcFfrOPcvx0Y041Iq
	EuLH2X2BnSFmyNiUqHooP9WWk6AvbLk41H/9ncW1g83251oirdC5XlyXjBMFcdVPjfvk
	U+7g==
X-Received: by 10.15.55.193 with SMTP id v41mr7646224eew.80.1391679996234;
	Thu, 06 Feb 2014 01:46:36 -0800 (PST)
Received: from localhost6.localdomain6 (38.248.broadband15.iol.cz.
	[90.182.248.38]) by mx.google.com with ESMTPSA id
	k41sm1460891een.19.2014.02.06.01.46.34 for <multiple recipients>
	(version=TLSv1.1 cipher=ECDHE-RSA-RC4-SHA bits=128/128);
	Thu, 06 Feb 2014 01:46:35 -0800 (PST)
From: David Jedelsky <david.jedelsky@gmail.com>
To: linux-media@vger.kernel.org
Cc: David Jedelsky <david.jedelsky@gmail.com>
Subject: [PATCH] [media] stb0899: Fix DVB-S2 support for TechniSat SkyStar 2
	HD CI USB ID 14f7:0002
Date: Thu,  6 Feb 2014 10:45:07 +0100
Message-Id: <1391679907-17876-1-git-send-email-david.jedelsky@gmail.com>
X-Mailer: git-send-email 1.7.10.4
Sender: linux-media-owner@vger.kernel.org
Precedence: bulk
List-ID: <linux-media.vger.kernel.org>
X-Mailing-List: linux-media@vger.kernel.org
X-PMX-Version: 6.0.0.2142326, Antispam-Engine: 2.7.2.2107409,
	Antispam-Data: 2014.2.6.93614
X-PMX-Spam: Gauge=IIIIIIII, Probability=8%, Report='
	FORGED_FROM_GMAIL 0.1, HTML_00_01 0.05, HTML_00_10 0.05,
	BODY_SIZE_3000_3999 0, BODY_SIZE_5000_LESS 0,
	BODY_SIZE_7000_LESS 0, DKIM_SIGNATURE 0, URI_ENDS_IN_HTML 0,
	__ANY_URI 0, __CP_URI_IN_BODY 0, __FRAUD_BODY_WEBMAIL 0,
	__FRAUD_WEBMAIL 0, __FRAUD_WEBMAIL_FROM 0, __FROM_GMAIL 0,
	__HAS_FROM 0, __HAS_MSGID 0, __HAS_X_MAILER 0,
	__HAS_X_MAILING_LIST 0, __MIME_TEXT_ONLY 0,
	__PHISH_SPEAR_STRUCTURE_1 0, __SANE_MSGID 0,
	__STOCK_PHRASE_7 0, __TO_MALFORMED_2 0, __TO_NO_NAME 0,
	__URI_NO_WWW 0, __URI_NS , __YOUTUBE_RCVD 0'

My TechniSat SkyStar 2 HD CI USB ID 14f7:0002 wasn't tuning DVB-S2 channels.
Investigation revealed that it doesn't read DVB-S2 registers out of stb0899.
Comparison with usb trafic of the Windows driver showed the correct
communication scheme. This patch implements it.
The question is, whether the changed communication doesn't break other devices.
But the read part of patched _stb0899_read_s2reg routine is now functinally
same as (not changed) stb0899_read_regs which reads ordinrary DVB-S registers.
Giving high chance that it should work correctly on other devices too.

Signed-off-by: David Jedelsky <david.jedelsky@gmail.com>
---
 drivers/media/dvb-frontends/stb0899_drv.c |   44 +++++++++++------------------
 1 file changed, 17 insertions(+), 27 deletions(-)

diff --git a/drivers/media/dvb-frontends/stb0899_drv.c b/drivers/media/dvb-frontends/stb0899_drv.c
index 07cd5ea..3084cb2 100644
--- a/drivers/media/dvb-frontends/stb0899_drv.c
+++ b/drivers/media/dvb-frontends/stb0899_drv.c
@@ -305,19 +305,20 @@ u32 _stb0899_read_s2reg(struct stb0899_state *state,
 		.len	= 6
 	};
 
-	struct i2c_msg msg_1 = {
-		.addr	= state->config->demod_address,
-		.flags	= 0,
-		.buf	= buf_1,
-		.len	= 2
+	struct i2c_msg msg[] = {
+		{
+			.addr	= state->config->demod_address,
+			.flags	= 0,
+			.buf	= buf_1,
+			.len	= 2
+		}, {
+			.addr	= state->config->demod_address,
+			.flags	= I2C_M_RD,
+			.buf	= buf,
+			.len	= 4
+		}
 	};
 
-	struct i2c_msg msg_r = {
-		.addr	= state->config->demod_address,
-		.flags	= I2C_M_RD,
-		.buf	= buf,
-		.len	= 4
-	};
 
 	tmpaddr = stb0899_reg_offset & 0xff00;
 	if (!(stb0899_reg_offset & 0x8))
@@ -326,6 +327,7 @@ u32 _stb0899_read_s2reg(struct stb0899_state *state,
 	buf_1[0] = GETBYTE(tmpaddr, BYTE1);
 	buf_1[1] = GETBYTE(tmpaddr, BYTE0);
 
+	/* Write address	*/
 	status = i2c_transfer(state->i2c, &msg_0, 1);
 	if (status < 1) {
 		if (status != -ERESTARTSYS)
@@ -336,28 +338,16 @@ u32 _stb0899_read_s2reg(struct stb0899_state *state,
 	}
 
 	/* Dummy	*/
-	status = i2c_transfer(state->i2c, &msg_1, 1);
-	if (status < 1)
-		goto err;
-
-	status = i2c_transfer(state->i2c, &msg_r, 1);
-	if (status < 1)
+	status = i2c_transfer(state->i2c, msg, 2);
+	if (status < 2)
 		goto err;
 
 	buf_1[0] = GETBYTE(stb0899_reg_offset, BYTE1);
 	buf_1[1] = GETBYTE(stb0899_reg_offset, BYTE0);
 
 	/* Actual	*/
-	status = i2c_transfer(state->i2c, &msg_1, 1);
-	if (status < 1) {
-		if (status != -ERESTARTSYS)
-			printk(KERN_ERR "%s ERR(2), Device=[0x%04x], Base address=[0x%08x], Offset=[0x%04x], Status=%d\n",
-			       __func__, stb0899_i2cdev, stb0899_base_addr, stb0899_reg_offset, status);
-		goto err;
-	}
-
-	status = i2c_transfer(state->i2c, &msg_r, 1);
-	if (status < 1) {
+	status = i2c_transfer(state->i2c, msg, 2);
+	if (status < 2) {
 		if (status != -ERESTARTSYS)
 			printk(KERN_ERR "%s ERR(3), Device=[0x%04x], Base address=[0x%08x], Offset=[0x%04x], Status=%d\n",
 			       __func__, stb0899_i2cdev, stb0899_base_addr, stb0899_reg_offset, status);
